### PR TITLE
Add strategy three

### LINKED ISSUE #15: Avoid default values on Match and Player
- Avoid default players while creating a Match instance
- Avoid default strategy while creating a  Player instance

### LINKED ISSUE #20: Add strategy three
Add a new strategy that looks for the column that is closest to win in a vertical line for a given color, as long as a four-in-a-row could be completed. Return first non full column otherwise.

### PR DESCRIPTION
**Pull request by apojomovsky**
Fixes #20 
Fixes #15

### CODE REVIEW

**Comment by apojomovsky** [comment]
@basicNew already addressed your comments, this is ready for a second pass (the first was during the class) :P

File: game/gameboard.py [hunk 1]
```diff
@@ -66,6 +68,45 @@ def retrieve_emptiest_column(self):
...
+    def _retrieve_consecutive_elements_from_array(self, array):
+        """Reads the consecutive elements from a given array
+        Args:
+            array: a numpy array
+        Returns
+            a list of elements conformed by: (element, number_of_repetitions)
+            element: object of the same kind as the one from the array
+            number_of_repetitions: the number of consecutive repetitions
+        """
+        return [(sum(1 for i in g), k) for k, g in groupby(array)]
```
Comment (basicNew) [inline game/gameboard.py]: Please, use meaningful variable names instead of `i`, `g` and `k`.

Comment (basicNew) [inline game/gameboard.py]: I also wonder: is `sum(1 for i in g)` necessary? Isn't that just `len(g)`?

File: game/match.py [hunk 2]
```diff
@@ -3,16 +3,17 @@
...
 class Match(object):
-    def __init__(self, player_one = Player('red'),
-                 player_two = Player('blue'), board = GameBoard()):
+    def __init__(self, player_one, player_two, board):
         self._board = board
         self._active_player = None
-        self._players = [player_one, player_two]
+        self._player_one = player_one
+        self._player_two = player_two
+        self._players = [self._player_one, self._player_two]
```
Comment (basicNew) [inline game/match.py]: you can define this as a getter that creates the array on-the-fly. No need for an instance var here.

File: game/strategy_three.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+       line for a given color, as long as a four-in-a-row could be completed.
+       Returns first non full column otherwise
+    """
+    def _can_complete_four(self, board, column_index, color):
+        same_color_on_top = board.count_same_color_on_top(column_index, color)
+        current_free_entries = board.count_free_entries_on_column(column_index)
+        return same_color_on_top + current_free_entries >= 4
+
+    def _get_column_closest_to_win(self, board, color):
+        current_max = [0, 0]
```
Comment (basicNew) [inline game/strategy_three.py]: It would be much more readable if you split this in two temporal vars (one with the index and another with the value)

File: game/strategy_three.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        for column_index in range(1, GameBoard.COLUMNSCOUNT + 1):
+            if self._can_complete_four(board, column_index, color):
+                same_color_on_top = board.count_same_color_on_top(column_index, color)
+                if same_color_on_top > current_max[1]:
+                    current_max[0] = column_index
+                    current_max[1] = same_color_on_top
+        if current_max[1] > 0:
+            return current_max[0]
+        return False
```
Comment (basicNew) [inline game/strategy_three.py]: I'm not a fan of returning, sometimes a number and sometimes a boolean. Given that you are asking for a column, if you can't find any, `None` would be better suited.

File: tests/test_gameboard.py [hunk 5]
```diff
@@ -146,6 +146,56 @@ def test_validate_matrix_with_invalid_matrix_wrong_structure(self):
             ['blue', 'blue',  'red', 'blue',  'red',   'blue',  None],
             ['red',   'red',  'red', 'blue',  'red',   'blue',  None]]))
 
+    def test_count_same_color_on_top_of_first_column(self):
```
Comment (basicNew) [inline tests/test_gameboard.py]: Is this the most simple example? I'm sure we discussed to start with the empty board, then with a board that has a single chip, then multiple chips with the same color and then reach this one.

File: tests/test_player.py [hunk 6]
```diff
@@ -2,27 +2,24 @@
...
 from game.gameboard import GameBoard
+from game.strategy_one import StrategyOne
+from mock import MagicMock
 
 class TestPlayer(unittest.TestCase):
     def setUp(self):
-        self.player_blue = Player('blue')
-        self.player_red = Player('red')
+        strategy_dummy = StrategyOne()
+        strategy_dummy.return_column = MagicMock(return_value = 1)
```
Comment (basicNew) [inline tests/test_player.py]: Nice!

Comment (basicNew) [inline tests/test_player.py]: As a note: mocks actually allows you to do much interesting things, but are out of the scope of what we are covering now.

File: tests/test_player.py [hunk 7]
```diff
@@ -2,27 +2,24 @@
...
-                    ['red',  'blue',  'red', 'blue',  'red', 'blue',  'red'],
-                    ['blue',  'red', 'blue',  'red',  'red',  'red',  'red']])
+                    [None,   None, None, None, None, None, None],
+                    [None,   None, None, None, None, None, None],
+                    [None,   None, None, None, None, None, None],
+                    [None,   None, None, None, None, None, None],
+                    [None,   None, None, None, None, None, None],
+                    ['blue', None, None, None, None, None, None]])
         self.player_blue.play(board)
         self.assertEqual(board, expected_board)
```
Comment (basicNew) [inline tests/test_player.py:27]: I think there are a couple more interesting cases here: 
- Returns index out of bounds (either 0 or 8).
- Returns index in the edges (either 1 or 7)
- Returns an index in the middle.
- Returns an index on a column that is already full.

File: tests/test_strategy_three.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+from game.player import Player
+from game.gameboard import GameBoard
+from game.strategy_three import StrategyThree
+
+class TestPlayer(unittest.TestCase):
+    def setUp(self):
+        self.strategy_three = StrategyThree()
+        self.color = 'blue'
+
+    def test_strategy_three_choses_leftmost_column_from_empty_board(self):
```
Comment (basicNew) [inline tests/test_strategy_three.py:12]: Comment here would be nice as why it chooses that column

File: tests/test_strategy_three.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+           return the first non full column.
+        """
+        board = GameBoard.from_matrix([
+                    [None,     None,   None,   None,   None,   None,   None],
+                    ['red',   'red', 'blue',  'red',  'red',  'red', 'blue'],
+                    ['red',  'blue',  'red',  'red', 'blue', 'blue', 'blue'],
+                    ['blue', 'blue',  'red', 'blue',  'red',  'red',  'red'],
+                    ['red',  'blue',  'red', 'blue',  'red', 'blue',  'red'],
+                    ['blue',  'red', 'blue',  'red',  'red',  'red',  'red']])
+        self.assertEqual(1, self.strategy_three.return_column(board, self.color))
```
Comment (basicNew) [inline tests/test_strategy_three.py:68]: These tests look much better!

**Review by basicNew — COMMENTED** [review]
Left some comments

File: game/gameboard.py (hunk 1 above)
Comment (apojomovsky) [reply game/gameboard.py]: Since `g` is an iterator, I think it's not possible to just get `len(g)`

File: game/match.py (hunk 2 above)
Comment (apojomovsky) [reply game/match.py]: Could you give me some more pointers about it? I've just read about getters but couldn't find any that could create an array on the fly

**Comment by apojomovsky** [comment]
Already addressed your comments @basicNew, this is ready for another pass

File: game/match.py (hunk 2 above)
Comment (basicNew) [reply game/match.py]: yup, the code would look like:

```
def players(self):
    return [self._player_one, self._player_two]
```

and in your code, instead of doing `self._players` do `self.players()`

File: tests/test_player.py (hunk 7 above)
Comment (basicNew) [reply tests/test_player.py:27]: Did you see the comment above? ^

Comment (apojomovsky) [reply tests/test_player.py:27]: Yes, sorry. I wanted to ask you something here. Since here I'm testing the method play() only, isn't that redundant if I test all those corner cases? I mean, shouldn't that behavior be tested for each strategy on its own test?

Comment (basicNew) [reply tests/test_player.py:27]: They are slightly different cases. What you want to test here is how the player reacts based on what  column the strategy returns. By being able to mock the strategy you can manage this as you please, hence being able to easily generate all the cases mentioned above. 

On the other hand, you can (and should) test the strategy without even creating a player; just set up the board and send the `return_column` message. But there you will be testing that the strategy actually does what you expect it to do (so, if it works ok, it should never return things like 0 or 8 :))

**Review by basicNew — APPROVED** [review]
